src/Umbraco.Core/Services/Implement/ContentService.cs

```diff
@@ -1333,36 +1451,29 @@ bool PublishCultures(IContent c)
             return results;
         }
 
+        // shouldPublish: a function determining whether the document has changes that need to be published
+        //  note - 'force' is handled by 'editing'
+        // publishValues: a function publishing values (using the appropriate PublishCulture calls)
         private PublishResult SaveAndPublishBranchOne(IScope scope, IContent document,
-            Func<IContent, bool> editing, Func<IContent, bool> publishValues,
-            bool checkPath,
-            List<IContent> publishedDocuments,
+            Func<IContent, HashSet<string>> shouldPublish, Func<IContent, HashSet<string>, bool> publishCultures,
+            bool isRoot,
+            ICollection<IContent> publishedDocuments,
             EventMessages evtMsgs, int userId)
         {
-            // if already published, and values haven't changed - i.e. not changing anything
-            // nothing to do - fixme - unless we *want* to bump dates?
-            if (document.Published && (editing == null || !editing(document)))
-                return new PublishResult(PublishResultType.SuccessAlready, evtMsgs, document);
-
-            // publish & check if values are valid
-            if (publishValues != null && !publishValues(document))
-                return new PublishResult(PublishResultType.FailedContentInvalid, evtMsgs, document);
+            // use 'shouldPublish' to determine whether there are changes to be published
+            // if the document has no changes to be published - nothing to
+            // for an invariant content, shouldPublish may contain "*" to indicate changes
+            var culturesToPublish = shouldPublish?.Invoke(document);
+            if (culturesToPublish == null || culturesToPublish.Count == 0)
+                return new PublishResult(PublishResultType.SuccessPublishAlready, evtMsgs, document);
```

Comment:
@zpqrtbnk I don't think this status is right, we will end up with a SuccessPublishAlready for all documents that are not published when force == false. When a document is just going to be ignored because it doesn't have a published version and force == false, then we just shouldn't include a result for that document , i'll see if i can update

Reply:
@zpqrtbnk unfortunately i'm not going to have time to resolve this today but it seems like in order to do this, the `shouldPublish` callback needs to return 2 things: bool includeInPublishing, HashSet<string> culturesToPublish or something along these lines. When `!force && !root && !culturePublished (or !published for invariant), then it should simply not be included in the branch publishing and not include a result type. I have this working with the callback returning a tuple, but then it's really ugly because this bleeds into the public API method with callbacks which we don't want.